Add configuration for Bootstrap 5

janko opened the pull request:
It's significantly simpler than the Bootstrap 3 configuration, partly due to changes in Bootstrap 5, partly due to greater code reuse, and partly due to probably not support every use case.

* checkboxes & radio buttons
* labels (supports floating labels)
* text-based fields (supports readonly plain text fields)
* validation errors (supports tooltips)
* primary submit button (supports changing style)
* select boxes
* range and color inputs

```rb
Forme.form @article, { action: "/articles", method: :post } do |f|
  f.input :title, wrapper_attr: { class: "mb-3" }, help: "Make it meaningful"
  f.input :body, as: :textarea, wrapper_attr: { class: "mb-3" }
  f.button value: "Create Article", class: "btn-success"
end
```
```html
<form action="/articles" class="forme article" method="post">
  <div class="mb-3 string required">
    <label class="form-label label-before" for="article_title">Title</label>
    <input class="form-control" id="article_title" maxlength="255" name="article[title]" required="required" type="text"/>
    <div class="form-text">Make it meaningful</div>
  </div>
  <div class="mb-3 string required">
    <label class="form-label label-before" for="article_body">Body</label>
    <textarea class="form-control" id="article_body" maxlength="255" name="article[body]" required="required"/>
  </div>
  <input class="btn btn-success" type="submit" value="Create Article"/>
</form>
```

jeremyevans commented:
Thanks for the patch. I'll try to review this tomorrow.  It would also be good to update the demo site to show an example of this support.

janko commented:
Done ✅ The only thing I didn't like is having to pass `wrapper_attr: { class: "mb-3" }` for each field in the `f.inputs` call, I'm not sure if there is a better way to do it. I don't think it can be the default on the BS5 wrapper, because then I don't know how one would change it.

jeremyevans commented:
Thank you for working on this.

Looking at the demo site, `/bs5/album/basic/date` looks odd due to presumably block display for date/time inputs that should be inline or inline-block.

The bootstrap 5 links on the demo site index page still use `bs3` instead of `bs5`, so those are broken.

This drops the table and readonly displays in the switch from bs3 to bs5.  Is there a reason for that?

This modifies the demo site views to add a `mb-3` class in wrapper_attr, which goes against the point of the demo-site.  The point of the demo-site is to show that the same form code can result in completely different output depending on the configuration.  If the bs5 views need `mb-3` to look nice, that should be part of the bs5 support, it should not need to be listed explicitly in the views.  If you don't think it is appropriate for the default bs5 support, then can you make the appropriate changes to the demo-site CSS, or add a custom wrapper on the demo site that adds the `mb-3` class, and use that?  The form view files (e.g. album_basic.erb) should require no changes at all to support bs5.

Coverage-wise, this is almost fully covered (thanks!).  Only 2 missing branches, on lines 34 and 113 of bs5.rb.  Could you cover those branches?

janko commented:
> Looking at the demo site, /bs5/album/basic/date looks odd due to presumably block display for date/time inputs that should be inline or inline-block.

Yeah, this rendered as a table on BS3, which the new BS5 config doesn't yet support. I'll add it.

> The bootstrap 5 links on the demo site index page still use bs3 instead of bs5, so those are broken.

Oops, will fix it 👍🏻 

> This drops the table and readonly displays in the switch from bs3 to bs5. Is there a reason for that?

As mentioned above, I will add table support and bring it back. I also didn't implement a readonly formatter, because I didn't see the advantage over just passing `:readonly` and `:disabled` options to the input calls. I can add it.

> If the bs5 views need mb-3 to look nice, that should be part of the bs5 support, it should not need to be listed explicitly in the views. If you don't think it is appropriate for the default bs5 support, then can you make the appropriate changes to the demo-site CSS, or add a custom wrapper on the demo site that adds the mb-3 class, and use that?

I ideally wanted to use `mb-3` by default in BS5 config, but I don't know how the user would then override it, so I chose not to do it. OK, I will modify the CSS or add a custom wrapper 👍🏻

> Only 2 missing branches, on lines 34 and 113 of bs5.rb. Could you cover those branches?

Will do 👍🏻

janko commented:
I believe I addressed all the mentioned issues:

* added `bs5_table` and `bs5_readonly` formatters, and brought back the corresponding pages 
* tried adding a default `mb-3` margin to the `bs5` wrapper, to be overridden if the user passes `wrapper_attr: { class: "anything" }`, but I didn't find a way to distinguish between user-defined class and the one set by the `sequel` plugin, so I chose to update the demo site CSS in the end
* fixed the broken links
* added missing branch coverage

jeremyevans commented:
Awesome, thank you!  I'll try to get this merged tomorrow.

jeremyevans commented:
I merged this.  I tried to reduce duplication a little bit by using `Forme.attr_classes` more in the bs5 support, but I found that broke the tests as it changed the order of classes in the output.  I added a `Forme.attr_classes_after` method that allows the existing tests to pass.  Unless the order of classes matters, it would be best to switch all `Forme.attr_classes_after` usage to `Forme.attr_classes`, update the tests, then remove `Forme.attr_classes_after`.
